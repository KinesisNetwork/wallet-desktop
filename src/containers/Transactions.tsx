import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import * as React from 'react'
import { connect } from 'react-redux'

import { loadNextTransactionPage } from '@actions'
import { RootState } from '@store'

import { EmptyTransactions } from '@components/EmptyTransactions'
import { TransactionCard } from '@components/TransactionOperation'
import { ConnectionStage, TransactionOperationView } from '@types'

const mapStateToProps = ({
  transactions: { transactionOperations, isLoading },
  connections: { currentCurrency, currentStage },
}: RootState) => ({
  operations: transactionOperations,
  currency: currentCurrency,
  isLoading,
  isTestnet: currentStage === ConnectionStage.testnet,
})

const mapDispatchToProps = {
  loadNextTransactionPage,
}

type Props = ReturnType<typeof mapStateToProps> & typeof mapDispatchToProps

class TransactionsPresentation extends React.Component<Props> {
  componentDidMount() {
    window.addEventListener('scroll', this.handleScroll)
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScroll)
  }

  render() {
    const { isLoading, operations } = this.props

    if (isLoading) {
      return (
        <div className="has-text-centered">
          <span className="icon">
            <FontAwesomeIcon icon={['fal', 'spinner']} size="3x" />
          </span>
        </div>
      )
    }

    if (operations.length === 0) {
      return <EmptyTransactions />
    }

    return this.groupOperationsByDate()
  }

  handleScroll = () => {
    const { loadNextTransactionPage: loadNextPage } = this.props

    const { scrollHeight, clientHeight, scrollTop } = document.documentElement
    const shouldTriggerLoad = scrollTop === scrollHeight - clientHeight
    if (shouldTriggerLoad) {
      loadNextPage()
    }
  }

  groupOperationsByDate = () => (
    <div onScroll={this.handleScroll}>
      {Object.entries(this.groupOperations()).map(([date, ops]) => (
        <React.Fragment key={date}>
          <h1 className="subtitle">{date}</h1>
          {ops.map(op => (
            <TransactionCard
              key={op.operation.id}
              transactionWithOperation={op}
              currency={this.props.currency}
              isTestnet={this.props.isTestnet}
            />
          ))}
        </React.Fragment>
      ))}
    </div>
  )

  groupOperations = (): { [date: string]: TransactionOperationView[] } =>
    this.props.operations.reduce((acc, op) => {
      const date = op.date.toDateString()
      const value = acc[date] || []
      acc[date] = [...value, op]
      return acc
    }, {})
}

export const Transactions = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TransactionsPresentation)
