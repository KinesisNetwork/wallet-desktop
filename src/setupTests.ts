import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
