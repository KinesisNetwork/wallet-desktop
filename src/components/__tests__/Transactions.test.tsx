import { shallow } from 'enzyme';
import * as React from 'react'

import { Transactions } from '@components/Transactions'
import '../../setupTests'

describe('Transactions', () => {
  let props

  beforeEach(() => {
    props = {
      isLastPage: false,
      isLoading: false,
      loadNextTransactionPage: () => null,
      operations: [
        {
          date: '',
          fee: '0.00046',
          operation: {},
          isIncoming: true,
          source: '123'
        },
        {
          date: '',
          fee: '0.02251',
          operation: {},
          isIncoming: false,
          source: '456'
        }
      ]
    }
  })

  describe('Transactions', () => {
    it('renders correctly', () => {
      const wrapper = shallow(<Transactions {...props} />)

      expect(wrapper).toMatchSnapshot()
    })

    it('renders a primary heading and two TransactionCard components', () => {
      const wrapper = shallow(<Transactions {...props} />)

      const heading = wrapper.find('h1')

      expect(heading).toHaveLength(1)
      expect(heading.text()).toEqual('Transactions')
      expect(wrapper.find('TransactionCard')).toHaveLength(2)
    })

    it('calls the handleScroll method on scroll', () => {
      const handleScrollMock = jest.fn()
      const wrapper = shallow(<Transactions {...props} loadNextTransactionPage={handleScrollMock} />)

      wrapper.find('.scrollable').simulate(
        'scroll',
        {
          currentTarget: {
            scrollHeight: 0,
            scrollTop: 1,
            clientHeight: 0
          }
        }
      )

      expect(handleScrollMock).toHaveBeenCalled()
    })
  })
})
