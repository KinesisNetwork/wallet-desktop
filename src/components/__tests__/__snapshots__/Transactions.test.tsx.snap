// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions Transactions renders correctly 1`] = `
<Fragment>
  <h1
    className="sub-heading primary-font"
  >
    Transactions
  </h1>
  <div
    className="scrollable"
    onScroll={[Function]}
  >
    <TransactionCard
      key="0"
      transactionWithOperation={
        Object {
          "date": "",
          "fee": "0.00046",
          "isIncoming": true,
          "operation": Object {},
          "source": "123",
        }
      }
    />
    <TransactionCard
      key="1"
      transactionWithOperation={
        Object {
          "date": "",
          "fee": "0.02251",
          "isIncoming": false,
          "operation": Object {},
          "source": "456",
        }
      }
    />
  </div>
</Fragment>
`;
